Task completion

michael-smith-nz opened the pull request:
fixes #126

michael-smith-nz committed:
feature(task-complete): task complete now adds taskWork to db.

michael-smith-nz commented on tasks/containers/TaskWorker.js:
```diff
@@ -99,6 +100,24 @@ export default compose(
+      }
+    }
+
+    actions.taskWorks.create(taskWork)
```
Is this creation not async? I tried a callback and promise with no luck.

ahdinosaur replied:
we're using [redux-observable](https://redux-observable.js.org/) in [feathers-action](https://github.com/ahdinosaur/feathers-action), so it's async but currently doesn't return a promise it [returns a cancellation token](https://github.com/ahdinosaur/feathers-action-react/blob/master/index.js#L51-L56).

the pattern currently is you subscribe to everything you need upfront, so that when things change you're automatically subscribed. in this case, we want to be subscribed to all task works referenced by the current task plan, so that when we create a new task work we'll be immediately notified. this subscription stuff is the `query` and `shouldQueryAgain` functions in any container using `connectFeathers`.

i figured out how we could return a promise though: if we implement a [ready action](https://github.com/ahdinosaur/feathers-action/commit/24180dc526952c87121a012e1a7b66495ccfd16d) for all methods, then we could fire off the action and resolve or reject a promise based on ready or error. but i haven't done this, i've only recently had the idea.

michael-smith-nz committed:
redirect once taskworker is created

depends on ahdinosaur/feathers-action/#20

michael-smith-nz committed:
add taskWork to taskPlan getter

michael-smith-nz commented on tasks/containers/TaskWorker.js:
```diff
@@ -79,14 +79,32 @@ export default compose(
 
```
I'm getting an infinite `taskWork` request here because this is always returning true and I'm  struggling to find the reason why.

@ahdinosaur @iainkirkpatrick If either of you two gets a chance to look at this that'd be rad as I'm feeling blocked. Otherwise, coming back to it with fresh eyes tomorrow might be enough.

This branch also depends on this file to be changed https://github.com/ahdinosaur/feathers-action/pull/20/files to work

ahdinosaur replied:
hmmmm...

iainkirkpatrick replied:
which part always returns true @NotThatSmith ? if you're online in the next 40 mins or so i can take a look at it with you

michael-smith-nz replied:
@iainkirkpatrick Woops, that preview didn't work. This line https://github.com/root-systems/cobuy/blob/feature/task-completion/tasks/containers/TaskWorker.js#L78

Turns out https://github.com/ahdinosaur/feathers-action/pull/20 Didn't work how I intended anyway so might have to rethink things.

ahdinosaur replied:
@NotThatSmith in case you aren't already aware (i wasn't for a while): as a general debugging tip for this situation, i recommend using `debugger` (add a line with just `debugger`). it's like `console.log('value', value)`, but better since it halts execution and you can see all the local variables at the time. 🔍  then you can halt and play it for every time it reaches the questionable moment and see what is happening there.

ahdinosaur replied:
oh shit, i might have figured this out.

this line is bad: https://github.com/root-systems/cobuy/blob/feature/task-completion/tasks/containers/TaskWorker.js#L139

it should be `gte(__, 1)`, where `__` is a [special placeholder](http://ramdajs.com/0.21.0/docs/#__) when you need to curry in a particular order. the reason is because [`gte`](http://ramdajs.com/0.21.0/docs/#gte) expects arguments in `x > y` order, which means we are currying `x` as `1`, so it only works when it shouldn't work. (oops!)

another tip that helped me here (i found this while working on #141, made the same mistake a second time) was using [`tap`](http://ramdajs.com/0.21.0/docs/#tap) inside a ramda `pipe` like this.

michael-smith-nz replied:
Yah! That works, thanks @ahdinosaur! 😀

michael-smith-nz committed:
fix infinite query bug

michael-smith-nz committed:
fix false task complete bug

michael-smith-nz committed:
fix routing not being cleared on request

michael-smith-nz committed:
use fork of feather-action

can use main package once https://github.com/ahdinosaur/feathers-action/pull/20 has been merged

iainkirkpatrick reviewed (approved):
looks rad to me - tasks complete when i finish em - might be worth @ahdinosaur checking the PR on `feathers-action`, though we can always fix that later

iainkirkpatrick commented:
merging now `feathers-action` PR 21 has been merged
